File: BeomMin_13/week4/bst.c
#include "bst.h"

void insert(int key, int data) {
	node* newNode = make_node(key, data);
	if (search(newNode->key) != NULL) {
		printf("%d is exist\n", newNode->key);
		exit(1);
	}

	if (root == NULL) {
		root = newNode;
	}
	else {
		node* temp = root;
		search_insert_space(temp, newNode);
	}
}

void delete(int key) {
	if(search(key) == NULL) {
		printf("%d is not exist\n",key);
		exit(1);
	}
	else {
		node* delNode = search(key);
		node* repNode;
		//node* lefNode;
		if (HaveZeroChild(delNode)) {
			zero_deletion(delNode);
		}
		else if (HaveTwoChild(delNode)) {	
			repNode = delNode->left_child;
			repNode = search_replace_node(repNode);			
			
			//lefNode = left_node_of_replace_node(repNode, lefNode);

			if (delNode == root) {
				two_deletion_root(delNode, repNode);			
			}
			else if (delNode->left_child == repNode || delNode->right_child == repNode ) {
				two_deletion_replace_node(delNode, repNode);
			}
			else {
				two_deletion_middle(delNode, repNode);
			}
		}
		else {
			if (delNode->left_child != NULL) {
				one_deletion_left(delNode);
			}
			else {
				one_deletion_right(delNode);
			}
		}
		init_delete_node(delNode);
	}
}

node* search(int key) {
	node* search_node = root;
	if (root == NULL) {
		return NULL;
	}
	else {
		while (search_node != NULL) {
			if (search_node->key == key) {
				return search_node;
			}
			else if(search_node->key > key) {
				MoveLeft(search_node);
			}
			else {
				MoveRight(search_node);
			}
		}
		return NULL;
	}
}	
	
		
void inorder(node* node) {
	if(node->left_child != NULL) {
		inorder(node->left_child);
	}
	printf("%d ", node->key);
	if(node->right_child != NULL) {
		inorder(node->right_child);
	}
}

node* get_root() {
	return root;
}
	

File: yohan_13/yohan_13/BST.java
package BST;
public class BST {
	private Node root;
	
	private class Node{
		private int key;
		private Node left_child;
		private Node right_child;
		private Node parent;
		Node(int key){
			this.key = key;
		}
	}
	
	public void insert(int key){
		if(this.search(key) != null){
			System.out.println(key + " is already in the tree");
			return;
		}
		Node newNode = new Node(key);	// assume the data is same as key
		//////////////////////////////////// todo
		if(root == null){
			root = newNode;
		}else{
			Node current = root;
			Node parent = current;
			while(true){
				parent = current;
				if(newNode.key<current.key){
				current = current.left_child;
				if(current == null){
					parent.left_child = newNode;
					return;
				}
			}else{
				current= current.right_child;
				if(current == null){
					parent.right_child = newNode;
					return;
					}
				}
			}
		}
	}
	public void delete(int key){
		//////////////////////////////////// todo
		 	Node current = root;
	        Node parent = current;

	        while(current.key != key){
	            if (current == null)        // 키 값을 갖는 노드 없음
	            	return;
	            parent = current;
	            if (key < current.key)
	                current = current.left_child;
	            else
	                current = current.right_child;
	        }

	        // 자식이 없을 때
	        if (current.left_child == null && current.right_child == null)
	        {
	            if (current == root)        // 트리 삭제
	                root = null;
	            else if (current == parent.left_child)
	                parent.left_child = null;
	            else
	                parent.right_child = null;
	        }

	        // 자식이 하나일 때
	        else if (current.right_child == null)        // 왼쪽자식만 존재
	        {
	            if (current == root)
	                root = current.left_child;
	            else if (current == parent.left_child)
	                parent.left_child = current.left_child;
	            else
	                parent.right_child = current.left_child;
	        }

	        else if (current.left_child == null)            // 오른쪽자식만 존재
	        {
	            if (current == root)
	                root = current.right_child;
	            else if (current == parent.left_child)
	                parent.left_child = current.right_child;
	            else
	                parent.right_child = current.right_child;
	        }

	        // 자식이 둘일 때
	        else{
	            // 삭제노드의 후보노드 찾기
	            Node candidate = getCandidate(current);

	            if (current == root)
	                root = candidate;
	            else if (current == parent.left_child)
	                parent.left_child = candidate;
	            else
	                parent.right_child = candidate;
	            
	            candidate.right_child = current.right_child;
	        }
	}
	
	private Node getCandidate(Node deleteNode)
    {
        Node candidateParent = deleteNode;
        Node candidate = candidateParent.left_child;

        // 삭제노드 오른쪽 자식의 왼쪽 자손 찾기
        while (candidate.right_child != null){
            candidateParent = candidate;
            candidate = candidate.right_child;
        }
        
        if (candidate != deleteNode.left_child)
        {
            candidateParent.right_child = candidate.left_child;
            candidate.left_child = deleteNode.left_child;
        }

        return candidate;
    }
	public Node search(int key){
		/////////////////////////// todo
		Node current = root;
		
		while(current!=null){
			if(current.key == key)
				return current;
			else if(key<current.key)
				current = current.left_child;
			else
				current = current.right_child;
			}
			return current;
		}
	
	public void printPreorder(Node node){
		
		if (node != null){
			System.out.printf("%d ",node.key);
            printPreorder(node.left_child);
            printPreorder(node.right_child);
        }
		
	}
	public void printInorder(Node node){
		if (node != null){
            printInorder(node.left_child);
            System.out.printf("%d ",node.key);	
            printInorder(node.right_child);
        }
	}
	public void printPostorder(Node node){
	
		 if (node != null){
	            printPostorder(node.left_child);
	            printPostorder(node.right_child);
	            System.out.printf("%d ",node.key);
	        }
	}
	public Node getRoot(){
		return root;
	}
}
